src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,393 @@
+        cv::Mat mat(
+            1,
+            static_cast<int>(img_buf.size()),
+            CV_8U,
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+            static_cast<void *>(const_cast<std::byte *>(img_buf.data())));
+        switch (img_dims_[0]) {
+        case 1:
+            mode = cv::ImreadModes::IMREAD_GRAYSCALE;
+            type = CV_8UC1;
+            break;
+        case 3:
+            mode = cv::ImreadModes::IMREAD_COLOR;
+            type = CV_8UC3;
+            break;
+        case 4:
+            mode = cv::ImreadModes::IMREAD_UNCHANGED;
+            type = CV_8UC4;
+            break;
+        default:
+            throw std::invalid_argument{
+                fmt::format("Unsupported number of channels entered in the "
+                            "image_dimensions parameter: {0:n}",
+                            params_.image_dimensions[0])};
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins);
+        if (tmp.empty()) {
+            return nullptr;
+        }
+        if (params_.resize.has_value()) {
+            if (!resize(tmp, tmp, ins)) {
+                return nullptr;
+            }
+        }
+        if (params_.to_rgb && tmp.channels() != 1) {
+            try {
+                cv::cvtColor(tmp, tmp, cv::COLOR_BGR2RGB);
+            }
+            catch (cv::Exception &e) {
+                if (bbh_ != bad_batch_handling::skip) {
+                    auto msg =
+                        fmt::format("BGR2RGB operation for image: {0:n} "
+                                    "in data_store: {1:s} failed with the "
+                                    "following exception: {2:s}.",
+                                    ins.index(),
+                                    ins.get_data_store().id(),
+                                    e.what());
+                    if (bbh_ == bad_batch_handling::error) {
+                        throw std::runtime_error{msg};
+                    }
+                    logger::warn(msg);
+                }
+                // skip batch
+                return nullptr;
+            }
+        }
+        cv::Mat dst{img_dims_[1], img_dims_[2], type, ptr_img_vec};
+        if (!crop(tmp, dst, ins)) {
+            return nullptr;
+        }
+        ptr_img_vec += instance_offset;
+    }
+
+    auto ptr = wrap_cpu_array<data_type::uint8>(std::move(img_vec));
+    auto tsr =
+        make_intrusive<dense_tensor>(size_vector{batch_size,
+                                                 params_.image_dimensions[1],
+                                                 params_.image_dimensions[2],
+                                                 params_.image_dimensions[0]},
+                                     std::move(ptr));
+    return tsr;
+}
+
+cv::Mat
+image_reader::decode_image(cv::Mat const &buf,
+                           int mode,
+                           instance const &ins) const
+{
+    cv::Mat decoded_img = cv::imdecode(buf, mode);
```

Comment:
Shouldn't this call be wrapped in a try/catch block?

Reply:
I initially skipped that is documentation suggested that it will return an empty mat incase of invalid data or buffer (tested it as well). However, it may be safer to wrap it within try catch. updated.